tests/test_pyopener.py

```diff
@@ -272,3 +274,86 @@ def test_opener_fsspec_fs_tiff_threads_2():
             assert profile["driver"] == "GTiff"
             assert profile["count"] == 3
             assert src.read().shape == (3, 718, 791)
+
+
+def test_opener_multi_range_read():
+    """Test with Opener with multi-range-read method."""
+
+    class Opener(io.FileIO):
+        """Custom FileIO FS with `read_multi_range` method."""
+        def read_multi_range(
+            self,
+            nranges,
+            offsets,
+            sizes,
+        ):
+            warnings.warn("Using MultiRange Reads", UserWarning, stacklevel=2)
+            return [
+                self._read_range(offset, size) for (offset, size) in zip(offsets, sizes)
+            ]
+
+        def _read_range(self, offset, size):
+            _ = self.seek(offset)
+            return self.read(size)
+
+    class VSIOpener(_AbstractOpener):
+        """SubClass of _AbstractOpener with `read_multi_range` attribute."""
+        def __init__(self):
+            self._obj = Opener
+        def open(self, path, mode="r", **kwds):
+            return self._obj(path, mode=mode, **kwds)
+        def isfile(self, path):
+            return True
+        def isdir(self, path):
+            return False
+        def ls(self, path):
+            return []
+        def mtime(self, path):
+            return 0
+        def size(self, path):
+            with self._obj(path) as f:
+                return f.size()
+        def read_multi_range(self):
+            return True
+
+
+    with rasterio.open("tests/data/RGB.byte.tif", opener=VSIOpener()) as src:
+        profile = src.profile
+        assert profile["driver"] == "GTiff"
+        assert profile["count"] == 3
+        # Should emit a multi-range read
+        with pytest.warns(UserWarning, match="Using MultiRange Reads"):
+            _ = src.read()
+
+
+def test_opener_multi_range_read_err():
+    """Test with Opener without multi-range-read method. """
+
+    class VSIOpener(_AbstractOpener):
+        """SubClass of _AbstractOpener with `read_multi_range` attribute.
+
+        The class will have the read_multi_range attr, but the file handler won't have it.
+        """
+        def __init__(self):
+            self._obj = io.FileIO
+        def open(self, path, mode="r", **kwds):
+            return self._obj(path, mode=mode, **kwds)
+        def isfile(self, path):
+            return True
+        def isdir(self, path):
+            return False
+        def ls(self, path):
+            return []
+        def mtime(self, path):
+            return 0
+        def size(self, path):
+            with self._obj(path) as f:
+                return f.size()
+        def read_multi_range(self):
+            return True
```

Comment:
@vincentsarago should this implement multi range reading, even if only in a limited way, to demonstrate the feature?

Reply:
The real `read_multi_range` implementation is in the `opener.obj` directly because it's the method that will be called because we return `self._obj.open(**kwargs)` 

https://github.com/rasterio/rasterio/blob/dd1f9c1f31ff3051836a9b4828e09d4e779c97ff/tests/test_pyopener.py#L303-L304

the attribute in the wrapper class is just used as `hint`. 

We could maybe rename this as `has_optimized_read_multi_range`: https://github.com/OSGeo/gdal/blob/dd3396b45944799ed6492267c2754f26f635a7dc/port/cpl_vsil_plugin.cpp#L238-L246

Reply:
@vincentsarago 

We're talking about using a new kind of Python HTTP client as the opener, right? In another issue, I think we already established that filesystems on our computers don't support multiple range reads and don't benefit from them. But data transfer over HTTP may benefit from this.

I don't see any existing packages that provide this kind of feature, but if a server supports multiple range requests (AWS does not, for one) you could, in Python, make a request and take the [multi-range response](https://datatracker.ietf.org/doc/html/rfc9110#name-media-type-multipart-bytera), split it at its boundaries, and return a sequence of byte ranges. That's my mental model of how this feature would be used in practice. Is this close to yours?

To make an HTTP range request, you must specify range or ranges in the headers for a request. Thus, every `VSIFilesystemPluginReadMultiRangeCallback` function call must make a new request. Thus, the `void *pFile` argument isn't a file object, but an HTTP session-like object or a connector object that encapsulates a session object and a URL. I see that you get this, too, I'm just writing it down for others that might be reading along.

For that reason, I prefer `get_byte_ranges()` to `read_multi_range()` on the Python side. A new GET request must be made for each set of byte ranges. We're only supporting byte ranges and not other kinds of ranges, yes? And we're not in the standard file reading mode.

Assigning a function to `callbacks_struct.read_multi_range` only when appropriate makes sense and I think that's the most effective thing from a GDAL perspective. However, I would prefer to not check for a hint attribute. Instead, I think we should assign to the callbacks struct based on the class of the opener or the protocols it provides.

From a protocol perspective, `VSIFilesystemPluginReadMultiRangeCallback()` needs a pFile that provides `get_byte_ranges(offsets, sizes)`. And `VSIFilesystemPluginOpenCallback()` must return an object that provides that method as well as others.

Instead of `if hasattr(opener, "read_multi_range")`, we might try `isinstance(opener, MultiByteRangeAccessor)` where project define

```
from typing import runtime_checkable, Protocol

@runtime_checkable
class MultiByteRangeCapable(Protocol):
    def open(self, path: str, mode: str = "r", **kwds) ->  MultiByteRangeAccessor:
        ...

class MultiByteRangeAccessor(Protocol):
    def get_byte_ranges(self, offsets, sizes) -> list[bytes]:
        ...
```

@vincentsarago if I read the Python docs right, then Rasterio and vsifile can each include such as snippet and we can get interoperability without either project depending on the other.

Are you interested in trying this?

Reply:
> We're talking about using a new kind of Python HTTP client as the opener, right? In another issue, I think we already established that filesystems on our computers don't support multiple range reads and don't benefit from them. But data transfer over HTTP may benefit from this.

Yes, https://github.com/vincentsarago/vsifile

Basically this client tries to adopt GDAL optimization (mostly `read_multi_range`, used in the GTIFF driver) 

> I don't see any existing packages that provide this kind of feature, but if a server supports multiple range requests (AWS does not, for one) you could, in Python, make a request and take the [multi-range response](https://datatracker.ietf.org/doc/html/rfc9110#name-media-type-multipart-bytera), split it at its boundaries, and return a sequence of byte ranges. That's my mental model of how this feature would be used in practice. Is this close to yours?

🤦 Oh I think I mis interpreted what the `read_multi_range` was doing. My experience was: when not using the `opener`, gdal can emit `merged` request for adjacent block, when we add the `multi_range_read` callback,  gdal will merge the requests into one request.

I think it's what @rouault was mentioning in https://github.com/rasterio/rasterio/pull/2969#issuecomment-1830365788

> But in that code path, there is also some effort made to merge together byte ranges that are consecutives (or very close to each other) to also minimize the number of ranges requested.

That makes sense now 😅, I still feel that the rasterio `opener` option should support having py opener that have the `read_multi_range` method. 

> Assigning a function to callbacks_struct.read_multi_range only when appropriate makes sense and I think that's the most effective thing from a GDAL perspective. However, I would prefer to not check for a hint attribute. Instead, I think we should assign to the callbacks struct based on the class of the opener or the protocols it provides.

Absolutely, I was happy to see the callbackstruct creation moved in the `open` step, but because now we encapsulate the opener in a Abstract class, we don't have direct access to the opener class to check what protocol it supports. 

> Instead of if hasattr(opener, "read_multi_range"), we might try isinstance(opener, MultiByteRangeAccessor) where project define

that's sounds good to me even if I'm not quite sure how to implement this

Reply:
@vincentsarago I tried coming up with an example that uses `typing.Protocol` but found that the runtime checker doesn't really work in our case because return value type hints aren't available at runtime. So, I think we have to rely on abstract base classes. Here's an example of that:

```python
"""mine.py: defines types that you implement."""

from abc import ABC, abstractmethod


class MultiByteRangeAccessor(ABC):
    @abstractmethod
    def get_byte_ranges(self, offsets, sizes) -> list[bytes]:
        pass


class MultiByteRangeCapable(ABC):
    @abstractmethod
    def open(self, path: str, mode: str = "r", **kwds) ->  MultiByteRangeAccessor:
        pass


def myfunc(obj):
    return isinstance(obj, MultiByteRangeCapable)
```

```python
"""yours.py: your implementation. The module doesn't need to depend on mine.py."""


class YourThing:
    def open(self, path: str, mode: str = "r", **kwds):
        return self
    def get_byte_ranges(self, offsets, sizes):
        return [bytes("000"), bytes("111")]
```

```python
"""app.py: a program that brings mine.py and yours.py together."""

from mine import MultiByteRangeCapable, myfunc
from yours import YourThing

print(myfunc(YourThing()))
# Output: False

MultiByteRangeCapable.register(YourThing)
print(myfunc(YourThing()))
# Output: True
```

Does `MultiByteRangeCapable` seems like an okay interface to you?

Reply:
In https://github.com/Toblerity/Fiona/pull/1415 I propose `MultiByteRangeResourceContainer`.